### PR TITLE
[Dev Fix] [EE]  (2/2) Fix Sidekiq 7 specs [LOYAL-10414]

### PR DESCRIPTION
**Pull request by afifsohaili-ascenda**
https://kaligo.atlassian.net/browse/LOYAL-10414

# Background

- Specs failed when used with Sidekiq 7

# Design

- After investigation, it turns out it's due to Sidekiq 7 changing from using `redis` gem to `redis-client` gem.
- Our mock_redis implements `redis`'s interface, so when this change happen in 7, it breaks.
- The gem authors think it's not worth updating since:
  - It's easier than ever to run redis alongside ruby in CI nowadays, unlike a decade ago when the gem was originally created. See:
    - https://github.com/sds/mock_redis/issues/253#issuecomment-1384865261
    - https://github.com/sds/mock_redis/issues/253#issuecomment-1401195720
    - https://github.com/sds/mock_redis/issues/253#issuecomment-1509718734
- Ultimately, the dependency on mock_redis is removed and we run the spec locally and on Jenkins against a real local redis.
- Added a note on README to ensure devs are aware.

# Testing

- Ensure specs pass on Jenkins

### CODE REVIEW

**Review by sswander — COMMENTED** [review]
Just curious, does it work if we mock `Redis.new` like what Max figured out? [Kaligo/loyalty_engine@`6842458` (#5209)](https://github.com/Kaligo/loyalty_engine/pull/5209/commits/68424583154cc67fad9caf9b7b6c2455b8a23b48#diff-89eebfcbc0f14b6d989517837ca1e94fce4e2ce9a03233641cd936f2b8d2ed94R139)

I think if we can keep mocking redis in specs that would be nice, because it's more portable (i.e. can run in CI without redis) but if not possible then I think this is the next best alternative


EDIT: or maybe a more implicit question is "why did sidekiq 7 work for LE with mock_redis but not sidekiq_adhoc_job?"

**Comment by afifsohaili-ascenda** [comment]
@sswander had to crack open mockredis vs redis-client vs sidekiq and figured it out finally.

what I found is:

- sidekiq 7 switched from redis to redis-client gem
- sidekiq_adhoc_job calls `erb` ([here](https://github.com/sidekiq/sidekiq/blob/dddf20bf0a6decfa6e19b04611fcb3a281d009a3/lib/sidekiq/web/action.rb#L47)) which in turn call `sscan` in redis client ([here](https://github.com/sidekiq/sidekiq/blob/dddf20bf0a6decfa6e19b04611fcb3a281d009a3/lib/sidekiq/api.rb#L1125-L1138))
- the issue occurs because:
  - in redis client, `sscan` only requires the key as the param
  - in redis-rb, `sscan` requires the cursor too (unlike in `redis-client` where it's defaulted to 0 when not provided)
- this sscan is called to get the number of `queues` and `processes`, I assume to display the count of jobs inside each queues in sidekiq admin's top nav
  - this call to `sscan` is probably never hit when using sidekiq just as background job operations, hence why we don't require the same kind of code acrobatics here for `sscan` here in LE

Screenshot 2024-05-16 at 2 29 10 PM

I tried updating Jenkinsfile to not have to download redis docker image but I encountered an issue that I couldn't figure out that think is not important to spend too much time into. The important thing is the spec runs without redis now. Tested locally:

Screenshot 2024-05-16 at 2 37 07 PM

File: Gemfile.lock
Comment (sswander) [inline Gemfile.lock:1]: To confirm, this Gemfile.lock is intended yeah? I could have misheard during our verbal discussion but I recall something like this file is not intended to be in this PR

If intended please ignore what I said 🙇

**Review by sswander — APPROVED** [review]
Thanks for the explanation and finding the workaround! Gotcha sir 👍

File: Gemfile.lock
Comment (afifsohaili-ascenda) [reply Gemfile.lock:1]: yes sir this is intended

**Pull request merged**